crawler/crawl_emap.py

```diff
@@ -0,0 +1,130 @@
+import requests
+import configparser
+from typing import Callable
+import psycopg2
+import psycopg2.extras
+from crawl import get_db_connection
+import datetime
+from flask import current_app
+import random
+from time import sleep
+import sys
+CONFIG_FILE = "../api/external/electricitymap/electricitymap.ini"
+
+def exponential_backoff(max_retries: int = 3,
+                        base_delay_s: int = 1,
+                        should_retry: Callable[[Exception], bool] = lambda _: True):
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            retries = 0
+            while True:
+                try:
+                    result_func = func(*args, **kwargs)
+                    return result_func
+                except Exception as ex:
+                    current_app.logger.debug(f"Attempt {retries + 1} failed: {ex}")
+                    if retries < max_retries and should_retry(ex):
+                        delay = (base_delay_s * 2 ** retries + random.uniform(0, 1))
+                        current_app.logger.debug(f"Retrying in {delay:.2f} seconds...")
+                        sleep(delay)
+                        retries += 1
+                    else:
+                        raise
+        return wrapper
+    return decorator
+
+def get_auth_token():
+    try:
+        parser = configparser.ConfigParser()
+        config_filepath = CONFIG_FILE
+        parser.read(config_filepath)
+        return parser['auth']['token']
+    except Exception as ex:
+        raise ValueError(
+            f'Failed to retrieve watttime credentials: {ex}') from ex
+
+
+def prepare_insert_query(data):
+    carbon_intensity = data.get('carbonIntensity', 0.0)
+    if carbon_intensity is None:
+        carbon_intensity = 0.0
+
+    low_carbon_percentage = 0.00
+    renewable_percentage = 0.00
+
+    datetime = data['datetime']
+
+    zone_id = data['zone']
+
+    query = """
+    INSERT INTO EMapCarbonIntensity(DateTime, ZoneId, CarbonIntensity, LowCarbonPercentage, RenewablePercentage)
+    VALUES (%s, %s, %s, %s, %s)
+    ON CONFLICT (DateTime, ZoneId) DO UPDATE SET
+    CarbonIntensity = EXCLUDED.CarbonIntensity,
+    LowCarbonPercentage = EXCLUDED.LowCarbonPercentage,
+    RenewablePercentage = EXCLUDED.RenewablePercentage;
+    """
+
+    return query, (datetime, zone_id, carbon_intensity, low_carbon_percentage, renewable_percentage)
+
+@exponential_backoff(should_retry=lambda ex: ex.response.status_code == 429)
+def fetch(zone):
+    url = f"https://api-access.electricitymaps.com/free-tier/carbon-intensity/history?zone={zone}"
+    headers = {
+        "auth-token": get_auth_token()
+    }
+
+    history_response = requests.get(url, headers=headers)
+
+    history_response_json = history_response.json()
+    try:
+        assert history_response.ok, f"Error in EMap request for getting history according to zone ({history_response.status_code}): {history_response.text}"
```

Comment:
This is silently ignoring the error (not throwing exception or returning some kind of signal). You should call `history_response.raise_for_status()` like here: https://github.com/c3lab-net/energy-data/blob/fda34e4c80db8f6eb4d2e21942956e66d2d00e28/api/external/electricitymap/ba_from_loc.py#L24
, which will generate the correct exception that's passed to the should_retry function of the decorator.

Also, the error handling logic in this function is not right:
1. you shouldn't call .json() before checking status
2. even when the request failed (`.ok` is not True), you are still returning the json, and the caller has no way of knowing this request has failed and will proceed with the next piece of code `update()` with unknown data.

Reply:
A better way to arrange the try catch is to have that in the outer most loop that you want to stop and handle the failure. In this case it's the `fetch_and_update()` function. There you can catch the error (and stop normal execution), and print it to stderr. In this function, you can simply let `raise_for_status()` to throw an exception.

Reply:
fixed by imitating the ba_from_log.py to trigger it correctly.